### PR TITLE
feat: initial version

### PR DESCRIPTION
**Pull request by jonaslagoni**
### Description
Here are some of the features I include in the template:
- The ability to change the payload right before sending it and after receiving it through hooks
- The template will generate an Object oriented structure, meaning all the messages, channels and schemas will be generated separately using quicktype.
- The ability to promisify reply callbacks.
- The ability to generate a test client based on the mirrored version of the provided AsyncAPI file. 

### TODO
- [x] Create initial client
- [x] Create initial test client
- [x] Add event emitter to client
- [x] Add generate test client argument for optional test client
- [x] Create hooks for managing data processing i.e. encryption etc
- [x] Test `BeforeSendingData` and `RecievedData` hooks for changing the message
- [x] Finish promisify parameter for reply callbacks
- [x] Ensure all related security schemas are supported and how to integrate into the template
**The security schemas have in this version no influence on what is generated. I don't have an actual use-case where this is necessary so this will not be prioritized.**
- [x] Add template readme file for usage
- [x] Cleanup code for review

### TODO (maybe)
Waiting for https://github.com/asyncapi/generator/issues/216 to decide whether to continue include the following:
- [x] ~~Decide if `allOf`, `anyOf` and `oneOf` should be included in generating messages and schemas. If not, remove existing code~~
**Changed to let quicktype generate the schemas. `allOf` messages are not supported.** 
- [x] ~~Add enum schema generation~~
**Letting quicktype handle this.**

### Issues it fixes
- Solves https://github.com/asyncapi/generator/issues/151

### CODE REVIEW

**Comment by fmvilas** [comment]
Hey mate! Just a quick reminder that we're migrating the template structure. You should move everything that should or might be copied to the output folder to its own folder called "template". Also, .filters, .hooks, and .partials are now called filters, hooks, and partials respectively. And they should be placed outside the "template" folder, i.e. in the root folder of this repo, along with the package.json file, etc.

See this PR as an example: https://github.com/asyncapi/nodejs-template/pull/5

**Comment by jonaslagoni** [comment]
Update, gonna remove all generated tests (not test client) and add it as future feature for the template.

**Comment by jonaslagoni** [comment]
Waiting for https://github.com/asyncapi/generator/issues/216 since I do not want to put too much time into something that later might be discarded.

**Comment by jonaslagoni** [comment]
Quick update, there are two example [AsyncAPI documents](https://github.com/jonaslagoni/ts-nats-template/tree/feature/initial_version/test/docs/0.1.0)  with each [generated client in the repository](https://github.com/jonaslagoni/ts-nats-template/tree/feature/initial_version/test/expected_comparators/latest). There are still issues with Schema generation and messages, so it will not be fully working yet.

File: test/expected_comparators/latest/frontend/src/tests/testclient/index.ts
Comment (ColinSullivan1) [inline test/expected_comparators/latest/frontend/src/tests/testclient/index.ts:142]: I'd suggest adding a connect method with a credentials file - it's a popular way to connect. 

e.g. usage is `await connect({url: 'tls://localhost:4222', nkeyCreds: "/tmp/seed.txt" });`

File: partials/index/Request.njk
Comment (aricart) [inline partials/index/Request.njk:18]: I am actually contemplating allowing the nats client to simply use binary payloads. And have the client decide to decode/encode itself. Part of the problem is that for example, if you have a JSON client, the API requires that everything be valid JSON.

If the decoding of the payload is a function on the message, it becomes possible to let the subscriber/requestor to decide how to parse it back. This would simplify the client a bit, at the expense of some changes in the publisher.

File: partials/index/Standard.njk
Comment (aricart) [inline partials/index/Standard.njk:13]: the package.json needs to have a way of enforcing the client that you are expecting. Otherwise any kind of change such as the upcoming v2 will break your generator.

File: partials/index/Standard.njk
Comment (aricart) [inline partials/index/Standard.njk:162]: Depending on the library, the nonceSigner will be generated for you. For sure if you specify as creds file, the nonce is generated on master. On v2, the client generates the signer function for you as well if the JWT and nkey are provided

File: template/package.json.njk
Comment (aricart) [inline template/package.json.njk:42]: 👍

File: test/expected_comparators/latest/frontend/src/tests/testclient/index.ts
Comment (jonaslagoni) [reply test/expected_comparators/latest/frontend/src/tests/testclient/index.ts:142]: Great catch and good question! I am still wondering if I should add these connection wrappers for simplification or just provide the default `connect` function and let the developer handle how the client should be connected.

Or if it should provide the wrappers (maybe dependent on the defined [security schemas](https://www.asyncapi.com/docs/specifications/2.0.0/#securitySchemeObject) to ease the interaction such as `connectWithNkey`. 

Any suggestions on what would be the prefered way? 🤔

File: partials/channel/Reply.njk
Comment (jonaslagoni) [reply partials/channel/Reply.njk:70]: Good idea. I'll create my own exception instead 👍

**Comment by ColinSullivan1** [comment]
Looks good; we offered a few suggestions to help moving forward.  Thank you so much for the work here - it's highly appreciated!

We're excited to be included and to support the AsyncAPI ecosystem!

File: partials/channel/Request.njk
Comment (jonaslagoni) [reply partials/channel/Request.njk:13]: Great suggestion, I'll do that instead 👍

File: partials/index/Standard.njk
Comment (jonaslagoni) [reply partials/index/Standard.njk:13]: @aricart can you elaborate on the release date for v2? And can you imagine people would like to use v1 rather then v2 in any scenario in regards to this template?

File: partials/index/Request.njk
Comment (jonaslagoni) [reply partials/index/Request.njk:18]: Interesting, it definitely took me off guard when first using the client that you had to have a client for each of the three protocols (compared to other client implementations such as the official  .NET client).

File: test/expected_comparators/latest/frontend/src/tests/testclient/index.ts
Comment (ColinSullivan1) [reply test/expected_comparators/latest/frontend/src/tests/testclient/index.ts:142]: The three most common ways to connect are plain url (no auth), user/pass, and credentials.  I'll leave it up to you to decide what is best for users in terms of API footprint but for ease of use I'd suggest APIs for plain url and credentials.  Can users still establish connections on their own?  For new projects we're recommending using NATS creds to connect.

Comment (jonaslagoni) [reply test/expected_comparators/latest/frontend/src/tests/testclient/index.ts:142]: Okay, i'll add url (no auth), user/pass, and credentials as build in connection functions. And yes people can still connect using their own custom connection options in any way they want 👍

File: partials/channel/Reply.njk
Comment (jonaslagoni) [reply partials/channel/Reply.njk:70]: Introduced NatsTypescriptTemplateError

File: test/expected_comparators/latest/frontend/src/tests/testclient/index.ts
Comment (ColinSullivan1) [reply test/expected_comparators/latest/frontend/src/tests/testclient/index.ts:142]: Thanks - I appreciate it!  Those should cover 99% of NATS use cases.

**Comment by jonaslagoni** [comment]
So it's finally to a point where I am happy to getting the template reviewed. I have provided examples for how to use the template as well as trying to cover everything in the README for people to understand how to use it. 

I would assume this PR should be reviewed as at the same time as https://github.com/asyncapi/bindings/pull/6 since this template is build around these bindings.

File: package.json
Comment (jonaslagoni) [inline package.json:3]: What should the initial version be?

**Comment by jonaslagoni** [comment]
@ColinSullivan1 @aricart just pinging you guys incase you want to have a look at the final version😄

**Comment by jonaslagoni** [comment]
And ofcourse the last refactor broke it and I didnt check it first 🤦‍♂️ Should be all good now...

File: README.md
Comment (fmvilas) [inline README.md:245]: I don't understand this.

File: README.md
Comment (fmvilas) [inline README.md:248]:
```suggestion
The first hook always receives the message type as is. Any intermediary hooks can return any type of your choosing. The last hook is required to return the correct message type.
```

File: README.md
Comment (fmvilas) [inline README.md:251]:
```suggestion
The first hook always receives the message type as is. Any intermediary hooks can return any type of your choosing. The last hook is required to return a string representation of the message.
```

File: README.md
Comment (fmvilas) [inline README.md:241]: It took me a while to realize you're not talking about Generator hooks here 😄

File: examples/publish subscribe/bin/StreetlightController.js
Comment (fmvilas) [inline examples/publish subscribe/bin/StreetlightController.js:1]: Can you replace tabs with spaces here? Github is showing me tabs as 8 spaces long 😅

File: README.md
Comment (jonaslagoni) [reply README.md:11]: You right I'll change it to the current version.

File: README.md
Comment (jonaslagoni) [reply README.md:55]: Hmm... IMO there is no reason for this. When I mention wildcards I mean [NATS topic wildcards](https://docs.nats.io/nats-concepts/subjects#wildcards). Meaning they are interpreted exactly like that 🤔

File: README.md
Comment (jonaslagoni) [reply README.md:241]: Yea no, that is the generated client's hooks 😄

**Comment by jonaslagoni** [comment]
@fmvilas changed the README file to use 2 spaces instead of tabs.

File: README.md
Comment (jonaslagoni) [reply README.md:196]: Does the generator support this? 🤔

File: README.md
Comment (jonaslagoni) [reply README.md:197]: Same as above

File: README.md
Comment (jonaslagoni) [reply README.md]: This has now been changed. Should be easier to understand.

**Comment by jonaslagoni** [comment]
I have noticed that a problem with parameter realization, gonna need to fix that before PR approval.

**Comment by jonaslagoni** [comment]
Fixed.

File: package.json
Comment (jonaslagoni) [reply package.json:3]: Gonna keep 0.1.0 as initial version.

File: README.md
Comment (jonaslagoni) [reply README.md:196]: Blocked by https://github.com/asyncapi/generator/issues/414

File: package.json
Comment (jonaslagoni) [reply package.json:3]: staying on 0.1.0

**Comment by jonaslagoni** [comment]
@derberg i'll wait with the release until the generator reaches version 1.

**Comment by jonaslagoni** [comment]
I have added some integration tests which the template generates in order to test that both the test client and and client can talk to each other and gets the correct data.

That finishes the initial version IMO, so this should be the last review unless other changes are requested.

**Comment by derberg** [comment]
@jonaslagoni I think we can merge it as then improve afterwords. Just solve conflict and remove `.github/workflows/automerge-release-pr-bump.yml ` as it is there in the master already. 

Remeber that `feat!` means 1.0.0. but we release 0.1.0

**Comment by jonaslagoni** [comment]
@derberg good call, all done 👍

**Comment by derberg** [comment]
@jonaslagoni you did not remove `.github/workflows/automerge-release-pr-bump.yml`, its latest version is in master already https://github.com/asyncapi/ts-nats-template/blob/master/.github/workflows/automerge.yml

File: .github/workflows/pull-request-testing.yml
Comment (derberg) [inline .github/workflows/pull-request-testing.yml:7]: fyi, you can have multiple jobs per file, so you can have one workflow for testing on PR and two jobs in it, one for normal testing and another for integration. You do not have to change it, just letting you know in case you didn'

**Comment by jonaslagoni** [comment]
> @jonaslagoni you did not remove `.github/workflows/automerge-release-pr-bump.yml`, its latest version is in master already https://github.com/asyncapi/ts-nats-template/blob/master/.github/workflows/automerge.yml

Yea merged master into my PR. But I have removed them now.

**Review by derberg — APPROVED** [review]

**Comment by derberg** [comment]
@fmvilas as per my comment

> @jonaslagoni I think we can merge it as then improve afterwords

can you approve PR too?

**Review by fmvilas — APPROVED** [review]

**Comment by jonaslagoni** [comment]
@all-contributors please add @ColinSullivan1 review

**Comment by jonaslagoni** [comment]
@all-contributors please add @aricart review

**Comment by jonaslagoni** [comment]
@all-contributors please add @aricart review

**Pull request merged**
